src/slide-image-source.cc

```diff
@@ -0,0 +1,159 @@
+// -*- mode: c++; c-basic-offset: 4; indent-tabs-mode: nil; -*-
+// (c) 2021 Leonardo Romor <leonardo.romor@gmail.com>,
+//          Henner Zeller <h.zeller@acm.org>
+//
+// This program is free software; you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation version 2.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://gnu.org/licenses/gpl-2.0.txt>
+
+#include "slide-image-source.h"
+#include "framebuffer.h"
+#include "terminal-canvas.h"
+
+#include <cassert>
+#include <cstdio>
+#include <unistd.h>
+#include <openslide/openslide.h>
+
+#include <limits>
+#include <string>
+#include <utility>
+
+extern "C" {
+#  include <libswscale/swscale.h>
+}
+
+namespace timg {
+const char *SlideImageSource::VersionInfo() {
+    return openslide_get_version();
+}
+
+static void dummy_log(void *, int, const char *, va_list) {}
+
+// Scoped clean-up of c objects.
+struct ScopeGuard {
+    ScopeGuard(std::function<void()> f) : f_(f) {}
+    ~ScopeGuard() { f_(); }
+    std::function<void()> f_;
+};
+
+static inline bool invalid_dimensions(int64_t width, int64_t height) {
+    return (width < 0 || height < 0) ||
+        (width > std::numeric_limits<int>::max() ||
+         height > std::numeric_limits<int>::max());
+}
+
+bool SlideImageSource::LoadAndScale(const DisplayOptions &opts, int, int) {
+    options_ = opts;
+    if (opts.scroll_animation ||
+        filename() == "/dev/stdin" || filename() == "-") {
+        return false;  // Not dealing with these now.
+    }
+
+    // Check if this format is supported
+    openslide_t *osr = openslide_open(filename().c_str());
+    if (!osr)
+        return false;
+
+    // Automatically cleanup openslide context
+    ScopeGuard s([osr]() { openslide_close(osr); });
+
+    int64_t width, height;
+    int target_width;
+    int target_height;
+    double scaling;
+
+    // First check if there's a thumbnail with enough resolution.
+    const char* const* associated_image_names = openslide_get_associated_image_names(osr);
+    char const *target_associated_image = NULL;
+    while (*associated_image_names) {
+        const char *name = *associated_image_names;
+        openslide_get_associated_image_dimensions(osr, name, &width, &height);
+        if (invalid_dimensions(width, height))
+            return false;
+
+        if (!strcmp(name, "thumbnail")) {
```

Comment:
do you always want the thumbnail ? It sounds like it can be a small image, but if you have a wezterm or kitty terminal, these can display images likely 1000x1000 depending on your screen.

Is there a way to find the smallest image that is bigger than the requested size ? Or is there only thumbnail and full resolution ?

Reply:
(solved): thumbnails in slides are usually quite big (~1000x1000)